docs/proposals/component-registration.md

```diff
@@ -0,0 +1,409 @@
+  - Similar to Pods
+    - Listable, like most other top-level API resources
+    - Have both Spec and Status (filterable on both/either)
+    - Spec would have (optional) LivenessProbe and ReadinessProbe for determining the condition
+    - Describes a single instance
+    - Can be selected by labels, metadata, spec, and status
+    - Is namespaced (components would use a specific namespace, likely "kube-system")
+    - Status would have an array of [Conditions](../devel/api-conventions.md#typical-status-properties)
+    - Can be named with a name generator (similar to pods created by replication controllers) when instances of the same type don't already have unique IDs (generated name is returned in the create response)
+  - Similar to Endpoints (current External Endpoints usage)
+    - Spec would have address (ip or host) and ports (optionally named) to connect with an external service provider instance
+    - Manually added for non-component External Service provider instances (e.g. Oracle DB) that do not know about k8s (rather than adding Endpoints)
+  - Different from Endpoints
+    - Singular, instead of plural (use -List, namespacing, and labels for grouping)
+    - Includes both Ready & NotReady endpoints
+    - Supports multiple and future conditions (in an extensible Conditions array)
+2. Modify ServiceSpec.Selector to select from both Pods and ExternalServiceProviders
+3. Add a new ExternalServiceProvider Controller that handles probing ExternalServiceProviders
+  - Probe results update ExternalServiceProvider.Status.Conditions
+4. Modify Endpoints Controller to manage ExternalServiceProvider Endpoints
+  - Readiness from ExternalServiceProvider.Status.Conditions analogous to PodStatus.Conditions
+5. Modify the Primary Components to self-register ExternalServiceProviders (one per component instance) instead of an Endpoint
+  - Group ExternalServiceProviders (component instances) with labels (e.g. "type=component", "component-type=core")
+6. Modify the Addon Components' Pod and Service registration to use similar labels (e.g. "type=component", "component-type=addon")
+7. Register component Services (with ServiceSpec.Selector) as part of cluster deployment
+  - Services of new components can also be added at runtime, before any instances exist. Each instance would self-register a ExternalServiceProviders.
+8. Deprecate `/componentsstatuses` in favor of requesting a filtered list of ExternalServiceProviders
+  - ExternalServiceProviders are external-only, but have generic conditions (liveness and readiness)
+  - Endpoints are external + internal, but only have readiness (for now)
+
+
+#### API Types
+
+The following is a draft of how the API models might change.
+
+```
+type ExternalServiceProvider struct {
+	TypeMeta
+	ObjectMeta
+	Spec ExternalServiceProviderSpec
+	Status ExternalServiceProviderStatus
+}
+
+type ExternalServiceProviderSpec struct {
+	Host string
+	Ports []ExternalServiceProviderPort
+	LivenessProbe *Probe
```

Comment:
We use liveness to determine when a container should be killed and restarted. Since it cannot have that meaning in this scenario, I would prefer to use another name. I suggest `ReachabilityProbe`, which would produce a `Reachable` condition. From the description I saw elsewhere about using a TCP probe, it sounds like that's what you want.

Also, I don't want to gate this proposal on this issue, but FYI there have been requests for per-port readiness probes and conditions, which is effectively what we have internally: liveness per container, readiness per port. However, I was trying to keep the API simpler. In many/most cases, readiness per container would work. While we do record readiness per container, services currently use the pod-level readiness for all ports/containers. We could potentially do better in cases where the user specified container ports and used them in the probes.